runtime/src/bank/partitioned_epoch_rewards/mod.rs

```diff
@@ -11,13 +11,47 @@ use {
         partitioned_rewards::PartitionedEpochRewardsConfig, stake_rewards::StakeReward,
     },
     solana_sdk::{
-        account::AccountSharedData, clock::Slot, feature_set, pubkey::Pubkey,
-        reward_info::RewardInfo, stake::state::Delegation,
+        account::AccountSharedData,
+        account_utils::StateMut,
+        clock::Slot,
+        feature_set,
+        pubkey::Pubkey,
+        reward_info::RewardInfo,
+        stake::state::{Delegation, Stake, StakeStateV2},
     },
     solana_vote::vote_account::VoteAccounts,
     std::sync::Arc,
 };
 
+#[derive(AbiExample, Debug, Serialize, Deserialize, Clone, PartialEq)]
+struct PartitionedStakeReward {
+    /// Stake account address
+    pub stake_pubkey: Pubkey,
+    /// `Stake` state to be stored in account
+    pub stake: Stake,
+    /// RewardInfo for recording in the Bank on distribution. Most of these
+    /// fields are available on calculation, but RewardInfo::post_balance must
+    /// be updated based on current account state before recording.
+    pub stake_reward_info: RewardInfo,
+}
+
+impl PartitionedStakeReward {
+    fn maybe_from(stake_reward: &StakeReward) -> Option<Self> {
+        if let Ok(StakeStateV2::Stake(_meta, stake, _flags)) = stake_reward.stake_account.state() {
+            Some(Self {
+                stake_pubkey: stake_reward.stake_pubkey,
+                stake,
+                stake_reward_info: stake_reward.stake_reward_info,
+            })
+        } else {
+            None
+        }
+    }
```

Comment:
nit: is there a reason for this to be `maybe_from` instead of `TryFrom`? `TryFrom` would be more idiomatic, but maybe it makes more sense to use options for later work

Reply:
Ah right, that's a good point, feel free to leave it as is then, no worries